### PR TITLE
raftstore: always update read delegate's region

### PR DESCRIPTION
**Pull request by overvenus**
Signed-off-by: Neil Shen <overvenus@gmail.com>

## What have you changed? (mandatory)

Always update read delegate's region to avoid stale region info in the localreader. So read delegates always have up to data region when their peer become leader.

## What are the type of the changes? (mandatory)

- Bug fix (non-breaking change which fixes an issue)

## How has this PR been tested? (mandatory)

Unit tests.

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

## Refer to a related PR or issue link (optional)

#2639

### CODE REVIEW

File: src/raftstore/store/peer.rs [hunk 1]
```diff
@@ -539,7 +539,8 @@ impl Peer {
         }
         self.mut_store().set_region(region.clone());
         let progress = ReadProgress::region(region);
-        self.maybe_update_read_progress(progress);
+        // Always update read delegate's region.
```
Comment (Connor1996) [inline src/raftstore/store/peer.rs]: maybe you should add a comment for reason why we always update it.

File: src/raftstore/store/worker/read.rs [hunk 2]
```diff
@@ -358,6 +361,7 @@ impl<C: Sender<StoreMsg>> LocalReader<C> {
             Ok(RequestPolicy::ReadLocal) => Ok(Some(delegate)),
             // It can not handle other policies.
             Ok(_) => Ok(None),
+
```
Comment (Connor1996) [inline src/raftstore/store/worker/read.rs]: please remove blank line

File: src/raftstore/store/worker/read.rs [hunk 3]
```diff
@@ -103,16 +103,17 @@ impl ReadDelegate {
                 {
                     // Cache snapshot_time for remaining requests in the same batch.
                     *last_valid_ts = snapshot_time;
-                    if let Ok(mut resp) = executor.execute(req, &self.region) {
-                        // Leader can read local if and only if it is in lease.
-                        cmd_resp::bind_term(&mut resp.response, term);
-                        return Some(resp);
-                    }
+                    let mut resp = executor.execute(req, &self.region);
```
Comment (Connor1996) [inline src/raftstore/store/worker/read.rs]: if `check_region_epoch()` failed in `execute()`, now `handle_read()` will return `Some(resp with snap is None)` and before will return None, is that alright?

**Comment by BusyJay** [comment]
Please describe the bug you are trying to fix in the description.

File: src/raftstore/store/worker/read.rs (hunk 3 above)
Comment (overvenus) [reply src/raftstore/store/worker/read.rs]: Yes, it is okay, because we disable epoch check in localreader.

**Comment by BusyJay** [comment]
What if the request arrives before region is updated?

**Comment by overvenus** [comment]
It may be handled or rejected by lease, it depends to what kind of changes made to the region.

There are two categories of commands that changes region:
 1. conf changes, they have no effect to read delegates as long as theirs peers are leaders.
 2. split / merge
   - split: the request maybe handled, because peers of new region that in other store do not start election before theirs election timeout which is longer than the max leader lease
   - merge: we suspect lease right after prepare merge is committed, so the request is rejected.

File: src/raftstore/store/worker/read.rs (hunk 3 above)
Comment (BusyJay) [reply src/raftstore/store/worker/read.rs]: I think this will become a pitfall in the future.

File: src/raftstore/store/worker/read.rs [hunk 4]
```diff
@@ -325,10 +327,10 @@ impl<C: Sender<StoreMsg>> LocalReader<C> {
             }
             None => {
                 self.metrics.borrow_mut().rejected_by_no_region += 1;
+                debug!("rejectd by no region {}", region_id);
```
Comment (BusyJay) [inline src/raftstore/store/worker/read.rs]: s/rejectd/rejected/

File: src/raftstore/store/worker/read.rs [hunk 5]
```diff
@@ -350,6 +352,7 @@ impl<C: Sender<StoreMsg>> LocalReader<C> {
         if util::check_region_epoch(req, &delegate.region, false).is_err() {
             self.metrics.borrow_mut().rejected_by_epoch += 1;
             // Stale epoch, redirect it to raftstore to get the latest region.
+            debug!("{} rejectd by stale epoch", delegate.tag);
```
Comment (BusyJay) [inline src/raftstore/store/worker/read.rs]: Ditto.

File: src/raftstore/store/worker/read.rs (hunk 3 above)
Comment (overvenus) [reply src/raftstore/store/worker/read.rs]: Could you make an example?

**Comment by overvenus** [comment]
PTAL, thanks!

File: src/raftstore/store/worker/read.rs (hunk 3 above)
Comment (BusyJay) [reply src/raftstore/store/worker/read.rs]: It's not obvious for reader. If a reviewer misunderstood it, people could make the same mistake too when refactoring the code.

**Comment by hicqu** [comment]
LGTM.

**Pull request merged**